src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -100,18 +100,18 @@ public static async ValueTask<ImmutableArray<(Checksum, object)>> ReadDataAsync(
             // Use local pipe to avoid blocking the current thread on networking IO.
             var localPipe = new Pipe(PipeOptionsWithUnlimitedWriterBuffer);
 
-            Exception? exception = null;
-
             // start a task on a thread pool thread copying from the RPC pipe to a local pipe:
```

Comment:
I don't get why you're doing this. If the idea was to avoid blocking a thread to wait for more bytes to stream in, shouldn't you actually *await* this `copyTask` before consuming the `localPipe`? As it is, now you have a race between async reading and sync deserializing, which does nothing to limit either max memory consumed or time spent blocking a thread.

Reply:
If we wait for copyTask then we need to materialize the whole payload in memory before we start deserializing it. Why wouldn't async reading and sync deserializing limit max memory consumed if these run in parallel?

Reply:
In the linked issue I see copyTask completed:
![image](https://user-images.githubusercontent.com/41759/104078451-f3312b80-51d2-11eb-8d95-0fd88f8b8be2.png)

No data have been written to the local pipe:
![image](https://user-images.githubusercontent.com/41759/104078569-b154b500-51d3-11eb-98e1-dcbf209f731d.png)

No data are in the RPC pipe:
![image](https://user-images.githubusercontent.com/41759/104078592-d0534700-51d3-11eb-9d7c-082677b1b519.png)

Reply:
The other side of the RPC pipe is here:

http://sourceroslyn.io/#Microsoft.CodeAnalysis.Remote.Workspaces/SolutionAssetProvider.cs,36